core/src/main/java/com/datastax/oss/driver/internal/core/CassandraProtocolVersionRegistry.java

```diff
@@ -0,0 +1,180 @@
+
+  public CassandraProtocolVersionRegistry(String logPrefix) {
+    this(logPrefix, CoreProtocolVersion.values());
+  }
+
+  protected CassandraProtocolVersionRegistry(String logPrefix, ProtocolVersion[]... versionRanges) {
+    this.logPrefix = logPrefix;
+    this.versionsByCode = byCode(versionRanges);
+  }
+
+  @Override
+  public ProtocolVersion fromCode(int code) {
+    ProtocolVersion protocolVersion = versionsByCode.get(code);
+    if (protocolVersion == null) {
+      throw new IllegalArgumentException("Unknown protocol version code: " + code);
+    }
+    return protocolVersion;
+  }
+
+  @Override
+  public ProtocolVersion fromName(String name) {
+    for (ProtocolVersion version : versionsByCode.values()) {
+      if (version.name().equals(name)) {
+        return version;
+      }
+    }
+    throw new IllegalArgumentException("Unknown protocol version name: " + name);
+  }
+
+  @Override
+  public ProtocolVersion highestNonBeta() {
+    ProtocolVersion highest = versionsByCode.lastEntry().getValue();
+    if (!highest.isBeta()) {
+      return highest;
+    } else {
+      return downgrade(highest)
+          .orElseThrow(() -> new AssertionError("There should be at least one non-beta version"));
+    }
+  }
+
+  @Override
+  public Optional<ProtocolVersion> downgrade(ProtocolVersion version) {
+    Map.Entry<Integer, ProtocolVersion> previousEntry =
+        versionsByCode.lowerEntry(version.getCode());
+    if (previousEntry == null) {
+      return Optional.empty();
+    } else {
+      ProtocolVersion previousVersion = previousEntry.getValue();
+      // Beta versions are skipped during negotiation
+      return (previousVersion.isBeta()) ? downgrade(previousVersion) : Optional.of(previousVersion);
+    }
+  }
+
+  @Override
+  public ProtocolVersion highestCommon(Collection<Node> nodes) {
+    if (nodes == null || nodes.isEmpty()) {
+      throw new IllegalArgumentException("Expected at least one node");
+    }
+
+    SortedSet<CoreProtocolVersion> candidates = new TreeSet<>();
+
+    for (CoreProtocolVersion version : CoreProtocolVersion.values()) {
+      // Beta versions always need to be forced, and we only call this method if the version
+      // wasn't forced
+      if (!version.isBeta()) {
+        candidates.add(version);
+      }
+    }
+
+    // The C*<=>protocol mapping is hardcoded in the code below, I don't see a need to be more
+    // sophisticated right now.
+    for (Node node : nodes) {
+      CassandraVersion cassandraVersion = node.getCassandraVersion();
+      if (cassandraVersion == null) {
+        LOG.warn(
+            "[{}] Node {} reports null Cassandra version, "
+                + "ignoring it from optimal protocol version computation",
+            logPrefix,
+            node.getConnectAddress());
+        continue;
+      }
+      cassandraVersion = cassandraVersion.nextStable();
+      if (cassandraVersion.compareTo(CASSANDRA_210) < 0) {
+        throw new IllegalArgumentException(
```

Comment:
Should we create a specific exception type for this case?  In previous driver we had `UnsupportedProtocolVersionException` for this case.